src/prommis/hydrogen_decrepitation/hydrogen_decrepitation_furnace.py

```diff
@@ -0,0 +1,1651 @@
+            self.impurity_ele_list,
+            doc="mass fraction of impurity element in solid feed stream",
+        )
+        def mass_frac_comp_impurity_ele_feed(b, t, i):
+            if i == "Fe":
+                return (
+                    b.flow_mol_comp_impurity_feed[t, "Nd2Fe14B"]
+                    * b.am_Fe
+                    * 14
+                    / b.flow_mass_impurity_feed[t]
+                )
+
+        # mass fraction of impurity metal in the product stream
+        @self.Constraint(
+            self.flowsheet().config.time,
+            self.impurity_ele_list,
+            doc="mass fraction of impurity element in solid product",
+        )
+        def mass_frac_comp_impurity_ele_product_eqn(b, t, i):
+            if i == "Nd":
+                return (
+                    b.mass_frac_comp_impurity_ele_product[t, i]
+                    * b.flow_mass_product_recovered[t]
+                    == b.flow_mol_comp_product_recovered[t, "Nd2Fe14B"] * b.am_Nd * 2
+                )
+            elif i == "Fe":
+                return (
+                    b.mass_frac_comp_impurity_ele_product[t, i]
+                    * b.flow_mass_product_recovered[t]
+                    == b.flow_mol_comp_product_recovered[t, "Nd2Fe14B"] * b.am_Fe * 14
+                )
+            elif i == "B":
+                return (
+                    b.mass_frac_comp_impurity_ele_product[t, i]
+                    * b.flow_mass_product_recovered[t]
+                    == b.flow_mol_comp_product_recovered[t, "Nd2Fe14B"] * b.am_B
+                )
+
+        # solid product with 13 species defined in CoalRefuseParameters
+        @self.Constraint(
+            self.flowsheet().config.time,
+            self.config.solid_property_package.component_list,
+            doc="component flow of outlet solid stream",
+        )
+        def solid_outlet_comp_eqn(b, t, i):
+            if i == "Nd" or i == "Nd2Fe14B":
+                return (
+                    b.solid_out[t].mass_frac_comp[i]
+                    * pyunits.convert(
+                        b.solid_out[t].flow_mass, to_units=pyunits.kg / pyunits.second
+                    )
+                    == b.flow_mol_comp_product_recovered[t, i]
+                    * b.config.solid_property_package.mw[i]
+                )
+            else:
+                pass
+
+    def _make_energy_balance(self):
+        # molar enthalpy of impurity minerals in solid feed
+        @self.Expression(
+            self.flowsheet().config.time,
+            self.impurity_list,
+            doc="molar enthalpy of individual impurity minerals",
+        )
+        def enth_mol_comp_impurity_feed(b, t, i):
+            return (
+                b.enth0_comp_impurity[i]
+                + b.cp0_comp_impurity[i] * (b.temp_feed[t] - b.temp_ref)
+                + 0.5
+                * b.cp1_comp_impurity[i]
+                * (b.temp_feed[t] * b.temp_feed[t] - b.temp_ref * b.temp_ref)
```

Comment:
Why are you expressing `x ** 2` as `x * x` here? Doing that is more efficient in low level languages because we're avoiding a call to the `pow` function, which is calculated using `exp` and `log`, but that is many levels removed from how Pyomo builds expression trees.